change behavior of environment get and __getitem__

adqm opened the pull request:
I believe this should fix the issue from #610, by changing the way `__getitem__` and `get` work for `__xonsh_env__`.

scopatz commented:
Hi @adqm, Thanks for putting this in.  This doesn't seem to have the right containment behaviour:

```console
scopatz@localhost ~ $ 'AUTO_SUGGEST' in set(__xonsh_env__.keys())
False
scopatz@localhost ~ $ 'AUTO_SUGGEST' in __xonsh_env__
True
```

I think a simpler solution would be to just override the `__contains__()` method.  

Also this PR Causes lookup failures that should probably be None for not existent environment variables. 

```console
scopatz@localhost ~ $ $DOES_NOT_EXIST
Traceback (most recent call last):
  File "/home/scopatz/.local/lib/python3.5/site-packages/xonsh/base_shell.py", line 136, in default
    self.execer.exec(code, mode='single', glbs=self.ctx)  # no locals
  File "/home/scopatz/.local/lib/python3.5/site-packages/xonsh/execer.py", line 127, in exec
    return exec(code, glbs, locs)
  File "<xonsh-code>", line 1, in <module>
  File "/home/scopatz/.local/lib/python3.5/site-packages/xonsh/environ.py", line 304, in __getitem__
    raise KeyError()
KeyError
```

The fix is to either revert to the previous version or to change the parser to use the `get()` method for lookup rather than `__getitem__()`.  If we change to the `get()` method, the logic for finding the default should probably be moved to `get()`, allowing `__getitem__()` to fail if the key is not present.

scopatz commented:
I'm up for whatever you think is best ;)

adqm commented:
> This doesn't seem to have the right containment behaviour:

That should be fixed with this most recent push.  My thought was that I wanted anything in `_d` or in `defaults` to be treated as being `in` the environment (anything with a value).  If we want, I can add a separate check for things that were explicitly set by the user.

That said, I temporarily broke a bunch of other things.  I'll come back to this one.

adqm commented:
Okay, @scopatz, I think I made all the updates for what felt right to me (though I'm open to other ideas).

What I wanted:
* Keys in `_d` _and_ in `defaults` are both considered to be `in` the environment (feels like anything with a value should be considered to be `in`)
* In Python mode, lookups that fail raise the exception
* In subprocess mode, lookups that fail are replaced with an empty string

For example:
```
hartz@virtual1 ~ $ $FOO
KeyError: 'Unknown environment variable: $FOO'
hartz@virtual1 ~ $ echo $FOO

hartz@virtual1 ~ $ echo "$FOO"
$FOO
hartz@virtual1 ~ $ $FOO is None
KeyError: 'Unknown environment variable: $FOO'
hartz@virtual1 ~ $ $FOO = 7
hartz@virtual1 ~ $ $FOO
7
hartz@virtual1 ~ $ echo $FOO
7
hartz@virtual1 ~ $ echo "$FOO"
7
hartz@virtual1 ~ $ $FOO is None
False
```

scopatz commented on xonsh/environ.py:
```diff
@@ -286,45 +286,50 @@ def get_ensurer(self, key,
+
     def __iter__(self):
         yield from self._d
+        yield from self.defaults
```
Won't this duplicate keys that are in both `_d` and `defaults`?

scopatz replied:
namely, 

```console
scopatz@localhost ~ $ len(list(__xonsh_env__)) == len(set(__xonsh_env__))
False
scopatz@localhost ~ $ len(list(__xonsh_env__))
89
scopatz@localhost ~ $ len(set(__xonsh_env__))
78
```

scopatz commented:
Thanks @adqm - I think your take on this makes a lot of sense.  Other than the iter issue, could you maybe add a couple of lines to the tutorial docs on this behaviour and then also updated the changelog, please?  With those, I think that this is ready to go in!

adqm commented:
Okay, I think this should be ready for another look.  Since last time:
* fixed the issue with double-yielding keys
* added changelog entries
* added small bit to tutorial
* made sure `${}` for looking up environment variables had the same semantics (error in Python mode, empty string in subproc mode)

scopatz commented:
Thanks @adqm!
